src/lib/homestore_backend/replication_state_machine.cpp

```diff
@@ -266,38 +266,101 @@ int ReplicationStateMachine::read_snapshot_data(std::shared_ptr< homestore::snap
+        return;
+    }
+
+    // Check if the snapshot context is same as the current snapshot context.
+    // If not, drop the previous context and re-init a new one
+    if (m_snp_rcv_handler && m_snp_rcv_handler->snp_lsn_ != context->get_lsn()) { m_snp_rcv_handler.reset(nullptr); }
+
+    // Check message integrity
+    auto header = r_cast< const SyncMessageHeader* >(snp_data->blob.cbytes());
+    if (header->corrupted()) {
+        LOGE("corrupted message in write_snapshot_data, lsn:{}, obj_id {} shard {} batch {}", s->get_last_log_idx(),
+             obj_id.value, obj_id.shard_seq_num, obj_id.batch_id);
+        return;
+    }
+    if (auto payload_size = snp_data->blob.size() - sizeof(SyncMessageHeader); payload_size != header->payload_size) {
+        LOGE("payload size mismatch in write_snapshot_data {} != {}, lsn:{}, obj_id {} shard {} batch {}", payload_size,
+             header->payload_size, s->get_last_log_idx(), obj_id.value, obj_id.shard_seq_num, obj_id.batch_id);
+        return;
+    }
+    auto data_buf = snp_data->blob.cbytes() + sizeof(SyncMessageHeader);
+
+    if (obj_id.shard_seq_num == 0) {
+        // PG metadata & shard list message
+        RELEASE_ASSERT(obj_id.batch_id == 0, "Invalid obj_id");
+
+        // TODO: Reset all data of current PG - let's resync on a pristine base
+
+        auto pg_data = GetSizePrefixedResyncPGMetaData(data_buf);
+        m_snp_rcv_handler = std::make_unique< HSHomeObject::SnapshotReceiveHandler >(
+            *home_object_, pg_data->pg_id(), r_dev->group_id(), context->get_lsn(), r_dev);
+        m_snp_rcv_handler->process_pg_snapshot_data(*pg_data);
+        snp_data->offset =
+            objId(HSHomeObject::get_sequence_num_from_shard_id(m_snp_rcv_handler->get_next_shard()), 0).value;
+        LOGD("Write snapshot, processed PG data pg_id:{} {}", pg_data->pg_id(), log_suffix);
+        return;
+    }
+
+    RELEASE_ASSERT(m_snp_rcv_handler,
+                   "Snapshot receiver not initialized"); // Here we should have a valid snapshot receiver context
+    if (obj_id.batch_id == 0) {
+        // Shard metadata message
+        RELEASE_ASSERT(obj_id.shard_seq_num != 0, "Invalid obj_id");
+
+        auto shard_data = GetSizePrefixedResyncShardMetaData(data_buf);
+        auto ret = m_snp_rcv_handler->process_shard_snapshot_data(*shard_data);
+        if (ret) {
+            // Do not proceed, will request for resending the shard data
+            LOGE("Failed to process shard snapshot data lsn:{} obj_id {} shard {} batch {}, err {}",
+                 s->get_last_log_idx(), obj_id.value, obj_id.shard_seq_num, obj_id.batch_id, ret);
+        } else {
+            // Request for the next batch
+            snp_data->offset = objId(obj_id.shard_seq_num, 1).value;
+        }
+        LOGD("Write snapshot, processed shard data shard_seq_num:{} {}", obj_id.shard_seq_num, log_suffix);
+        return;
+    }
+
+    // Blob data message
+    RELEASE_ASSERT(obj_id.shard_seq_num ==
```

Comment:
should we add some logic to handle this case. for example ,skip this and resend the same request to leader.